src/aeso_ast_infer_types.erl

```diff
@@ -3292,49 +3317,59 @@ create_freshen_tvars() ->
 destroy_freshen_tvars() ->
     ets_delete(freshen_tvars).
 
-freshen_type(Ann, Type) ->
+freshen_type(Ann, Type, Ctx) ->
     create_freshen_tvars(),
-    Type1 = freshen(Ann, Type),
+    Type1 = freshen(Ann, Type, Ctx),
     destroy_freshen_tvars(),
     Type1.
 
 freshen(Type) ->
-    freshen(aeso_syntax:get_ann(Type), Type).
+    freshen(aeso_syntax:get_ann(Type), Type, none).
 
-freshen(Ann, {tvar, _, Name}) ->
+freshen(Ann, {tvar, _, Name}, _Ctx) ->
     NewT = case ets_lookup(freshen_tvars, Name) of
                []          -> fresh_uvar(Ann);
                [{Name, T}] -> T
            end,
     ets_insert(freshen_tvars, {Name, NewT}),
     NewT;
-freshen(Ann, {bytes_t, _, any}) ->
+freshen(Ann, {bytes_t, _, '_'}, Ctx) ->
     X = fresh_uvar(Ann),
-    add_constraint({is_bytes, X}),
+    add_constraint({is_bytes, Ctx, X}),
     X;
-freshen(Ann, T) when is_tuple(T) ->
-    list_to_tuple(freshen(Ann, tuple_to_list(T)));
-freshen(Ann, [A | B]) ->
-    [freshen(Ann, A) | freshen(Ann, B)];
-freshen(_, X) ->
+freshen(Ann, {bytes_t, _, fixed}, Ctx) ->
+    X = fresh_uvar(Ann),
+    add_constraint({is_fixed_bytes, Ctx, X}),
+    X;
+freshen(Ann, {fun_t, FAnn, NamedArgs, Args, Result}, Ctx) when is_list(Args) ->
+    {fun_t, FAnn, freshen(Ann, NamedArgs, Ctx),
+     [ freshen(Ann, Arg, [{arg, Ix} | Ctx]) || {Arg, Ix} <- lists:zip(Args, lists:seq(1, length(Args))) ],
```

Comment:
You know there is an amazing function lists:enumerate which was added to stdlib by a well known and respected compiler developer?

Reply:
Yes, but only in OTP-25 :-) we still support older versions I think!